### PR TITLE
Use stream for writing blocks & snapshots

### PR DESCRIPTION
**Pull request by liayoo**
Updated to use streams for writing blocks & snapshots (using [json-stream-stringify](https://github.com/Faleij/json-stream-stringify), but I think we can change this to [JSONStream](https://github.com/dominictarr/JSONStream) if we end up using the latter for read streams)

TODO: use streams for reading blocks & snapshots

### CODE REVIEW

File: common/file-util.js [hunk 1]
```diff
@@ -222,8 +260,19 @@ class FileUtil {
         }
       }
     } else {
-      // TODO(liayoo): Change this operation to be asynchronous
-      fs.writeFileSync(filePath, zlib.gzipSync(Buffer.from(JSON.stringify(snapshot))));
+      return new Promise((resolve) => {
+        new JsonStreamStringify(snapshot)
```
Comment (platfowner) [inline common/file-util.js:264]: Can JSONStream be used for writing snapshots as it's used for reading?

File: common/file-util.js [hunk 2]
```diff
@@ -124,8 +126,34 @@ class FileUtil {
     return _.endsWith(filePath, '.gz');
   }
 
-  // TODO(cshcomcom): Change to asynchronous.
-  static readCompressedJson(filePath) {
+  static async readCompressedJson(filePath) {
```
Comment (platfowner) [inline common/file-util.js:129]: as this code is assuming some internal structure of the data, should the name be more specific? e.g. readChunkedJsonAsync()?

File: node/index.js [hunk 3]
```diff
@@ -359,13 +361,26 @@ class BlockchainNode {
     const stateSnapshot = stateRoot.toStateSnapshot({ includeVersion: true });
     const radixSnapshot = stateRoot.toRadixSnapshot();
     const rootProofHash = stateRoot.getProofHash();
+    if (DevFlags.enableSnapshotChunks) {
```
Comment (platfowner) [inline node/index.js:364]: what about applying this to the block files as well? 
I'm working on converting objects to chunks and I think it's applicable to blocks as well.
What do you think?

File: common/file-util.js (hunk 1 above)
Comment (liayoo) [reply common/file-util.js:264]: I had some problems with writing with JSONStream.. I got it to work, but without gzipping the json file. Should we still try to use JSONStream for writing?

Comment (platfowner) [reply common/file-util.js:264]: oh, i see.
unless there is any compatibility issues, feel free to use any solution as they're just tools for writing and reading.

Comment (liayoo) [reply common/file-util.js]: done!!

File: common/file-util.js (hunk 2 above)
Comment (liayoo) [reply common/file-util.js]: updated~

File: node/index.js (hunk 3 above)
Comment (liayoo) [reply node/index.js]: Applied the logic only to the snapshots as discussed offline. PTAL~

File: blockchain/index.js [hunk 4]
```diff
@@ -178,7 +177,7 @@ class Blockchain {
     }
   }
 
-  addBlockToChainAndWriteToDisk(block, writeToDisk) {
+  async addBlockToChainAndWriteToDisk(block, writeToDisk) {
```
Comment (platfowner) [inline blockchain/index.js:180]: no need for this any more?

File: blockchain/index.js [hunk 5]
```diff
@@ -195,7 +194,7 @@ class Blockchain {
     this.addBlockToChain(block);
     this.updateNumberToBlockInfo(block);
     if (writeToDisk) {
-      this.writeBlock(block);
+      await this.writeBlock(block);
```
Comment (platfowner) [inline blockchain/index.js:197]: no need for this any more?

File: common/file-util.js [hunk 6]
```diff
@@ -134,7 +172,7 @@ class FileUtil {
     }
   }
 
-  static readJson(filePath) {
+  static async readJsonSync(filePath) {
```
Comment (platfowner) [inline common/file-util.js:175]: no need for 'async' any more?

File: blockchain/index.js [hunk 7]
```diff
@@ -38,7 +37,7 @@ class Blockchain {
   /**
    * Initializes the blockchain and returns whether there are block files to load.
    */
-  initBlockchain(isFirstNode, snapshot) {
+  async initBlockchain(isFirstNode, snapshot) {
```
Comment (platfowner) [inline blockchain/index.js:40]: no need for 'async' any more?

File: blockchain/index.js (hunk 5 above)
Comment (liayoo) [reply blockchain/index.js]: Ah... thanks for catching it! Removed.

File: common/file-util.js (hunk 6 above)
Comment (liayoo) [reply common/file-util.js]: ditto

File: blockchain/index.js (hunk 7 above)
Comment (liayoo) [reply blockchain/index.js:40]: ditto

File: blockchain/index.js (hunk 4 above)
Comment (liayoo) [reply blockchain/index.js]: ditto

File: blockchain/index.js (hunk 7 above)
Comment (liayoo) [reply blockchain/index.js]: ditto

File: test/unit/blockchain.test.js [hunk 8]
```diff
@@ -74,7 +74,7 @@ describe('Blockchain', () => {
           blockHash = block.hash;
         }
         blocks.push(block);
-        node1.bc.addBlockToChainAndWriteToDisk(block, true);
+        await node1.bc.addBlockToChainAndWriteToDisk(block, true);
```
Comment (platfowner) [inline test/unit/blockchain.test.js:77]: no need for 'await' any more?

File: test/unit/blockchain.test.js [hunk 9]
```diff
@@ -47,7 +47,7 @@ describe('Blockchain', () => {
   describe('with lots of blocks', () => {
     let blocks; let blockHash;
 
-    beforeEach(() => {
+    beforeEach(async () => {
```
Comment (platfowner) [inline test/unit/blockchain.test.js:50]: no need for 'async' any more?

File: common/file-util.js [hunk 10]
```diff
@@ -124,8 +127,43 @@ class FileUtil {
...
+          .on('end', () => {
+            logger.debug(`${LOG_HEADER} Reading done: ${JSON.stringify(chunks)}`);
+            resolve(ObjectUtil.fromChunks(chunks));
+          })
+          .on('error', (e) => {
+            logger.error(`${LOG_HEADER} Error while reading ${filePath}: ${e}`);
+            resolve(null);
+          });
+      });
+    } catch (err) {
```
Comment (platfowner) [inline common/file-util.js:152]: can leave some log messages with error.message?

File: common/file-util.js [hunk 11]
```diff
@@ -124,8 +127,43 @@ class FileUtil {
...
+      return null;
+    }
+  }
+
+  static readChunkedJsonSync(filePath) {
+    try {
+      const zippedFs = fs.readFileSync(filePath);
+      return ObjectUtil.fromChunks(JSON.parse(zlib.gunzipSync(zippedFs).toString()).docs);
+    } catch (err) {
```
Comment (platfowner) [inline common/file-util.js:161]: ditto?

File: common/file-util.js [hunk 12]
```diff
@@ -124,8 +127,43 @@ class FileUtil {
...
+      return null;
+    }
+  }
+
+  static readCompressedJsonSync(filePath) {
```
Comment (platfowner) [inline common/file-util.js:166]: can leave some log messages in error catching below?

File: common/file-util.js [hunk 13]
```diff
@@ -134,7 +172,7 @@ class FileUtil {
     }
   }
 
-  static readJson(filePath) {
+  static readJsonSync(filePath) {
```
Comment (platfowner) [inline common/file-util.js:175]: ditto for error catching?

**Review by platfowner — APPROVED** [review]
LGTM
Thanks for delivering this nice solution in such a short time!

File: common/file-util.js (hunk 10 above)
Comment (liayoo) [reply common/file-util.js:152]: done~

File: common/file-util.js (hunk 11 above)
Comment (liayoo) [reply common/file-util.js:163]: ditto

File: common/file-util.js (hunk 12 above)
Comment (liayoo) [reply common/file-util.js:169]: ditto

File: common/file-util.js (hunk 13 above)
Comment (liayoo) [reply common/file-util.js:180]: ditto

**Comment by liayoo** [comment]
Thanks for the thorough review! Now merging!!

**Pull request merged**